webpack.config.js

```diff
@@ -1,56 +1,53 @@
 const HtmlWebpackPlugin = require('html-webpack-plugin');
-const PORT = process.env.PORT || 3000;
+const { resolve } = require('path');
+const webpack = require('webpack');
 
 module.exports = {
-  devtool: 'cheap-module-eval-source-map',
   entry: [
     'react-hot-loader/patch',
-    `webpack-dev-server/client?http://localhost:${PORT}`,
+    'webpack-dev-server/client?http://localhost:8080',
     'webpack/hot/only-dev-server',
-    './src/index.js',
+    './index.js',
   ],
+  output: {
+    filename: 'bundle.js',
+    path: resolve(`${__dirname}/dist`),
+    publicPath: '/',
+  },
+  context: resolve(__dirname, 'src'),
+  devtool: 'cheap-module-eval-source-map',
+  devServer: {
+    hot: true,
+    contentBase: resolve(__dirname, 'dist'),
+    publicPath: '/'
+  },
   module: {
     loaders: [
       {
-        test: /\.js(x?)$/,
+        test: /\.js$/,
         exclude: /node_modules/,
-        loaders: ['babel'],
+        loaders: ['babel-loader'],
       }, {
         test: /\.(png|jpg|jpeg|gif|svg|woff|woff2|eot)$/,
-        loader: 'url',
+        loader: 'url-loader',
         query: {
           name: '[path][name].[ext]?[hash]',
           limit: 500,
         },
       }, {
         test: /\.scss$/,
         loaders: [
-          'style?sourceMap',
-          'css?modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]',
-          'postcss',
-          'sass?sourceMap',
+          'style-loader?sourceMap',
+          'css-loader?modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]',
+          'postcss-loader',
+          'sass-loader?sourceMap',
         ],
       },
     ],
   },
-  resolve: {
-    extensions: ['', '.js', '.jsx'],
-  },
-  output: {
-    filename: 'bundle.js',
-    path: `${__dirname}/dist`,
-    publicPath: '/',
-  },
-  devServer: {
-    port: PORT,
-  },
-  postcss: [
-    require('autoprefixer'),
-    require('precss'),
-  ],
   plugins: [
-    new HtmlWebpackPlugin({
-      template: './src/index.html',
-    }),
+    new HtmlWebpackPlugin({ template: './index.html' }),
+    new webpack.HotModuleReplacementPlugin(),
+    new webpack.NamedModulesPlugin(),
```

Comment:
Would you like to tell me why these two plugins need to used here? I'd like to learn more on that. Through my experience, HMRE has never reliably worked. :(